numpy/core/src/multiarray/ctors.c

```diff
@@ -958,6 +961,18 @@ PyArray_NewFromDescr(
         int nd, npy_intp const *dims, npy_intp const *strides, void *data,
         int flags, PyObject *obj)
 {
+#ifdef __GNUC__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wnonnull-compare"
+#endif
+    if (descr == NULL) {
+#ifdef __GNUC__
+#pragma GCC diagnostic pop
+#endif
```

Comment:
Not sure why I had to work so hard to do this check. The function is NUMPY_API, so it is on us to sanitize input. Why does gcc think descr can never be NULL?

Reply:
I don't agree that API automatically means we have to sanitize, I do not think the Python C-API does this.  But it _is_ fairly common for us to do, and in this case I think the pattern "relying" on it (at least theoretically) is common enough to do it.

The generated header include:
```
NPY_NO_EXPORT NPY_STEALS_REF_TO_ARG(2) NPY_GCC_NONNULL(1) NPY_GCC_NONNULL(2) PyObject * PyArray_NewFromDescr \
       (PyTypeObject *, PyArray_Descr *, int, npy_intp const *, npy_intp const *, void *, int, PyObject *);
```
So we mark this as NONNULL explicitly, actually.  This is done here:
https://github.com/numpy/numpy/blob/855b5ed82381c088a0dbf8e629927a7f337e2dbb/numpy/core/code_generators/numpy_api.py#L140

Reply:
Maybe we can get away with not sanitizing inputs (I think best practices is always to be liberal with input and strict with output), but using NonNull on API interfaces seems to me to be an anti-pattern. I will open an issue.

Reply:
Done in #20971

Reply:
No idea about the merit of having this in the public API.  I think we should just change it to `NonNull([1])` for the sake of this PR, though?

Reply:
If we do we can only put it into 1.23

Reply:
Hmmm, isn't that fully ABI/API compatible, since it is just an attribute (and relaxing an attribute at that)?

Reply:
Hmm, lets discuss tomorrow or so.  I thought we may not have to put any error at all (anywhere), because passing in NULL should only happen if the `NULL` exists due to an error being already in-progress.  And if that is not true, Python will set a `SystemError`...

Reply:
Let's leave that for #20971. For this PR, I pushed the check one level down.

Reply:
I removed the check.